framework/include/neml2/userobjects/NEML2FEInterpolation.h

```diff
@@ -0,0 +1,176 @@
+//* This file is part of the MOOSE framework
+//* https://mooseframework.inl.gov
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include <vector>
+#ifdef NEML2_ENABLED
+
+// system includes
+#include <unordered_map>
+
+// libmesh includes
+#include "MooseTypes.h"
+#include "libmesh/fe_type.h"
+#include "libmesh/id_types.h"
+
+// MOOSE includes
+#include "ElementUserObject.h"
+#include "NEML2Assembly.h"
+#include "MooseVariableFE.h"
+
+// NEML2 includes
+#include "neml2/tensors/Tensor.h"
+
+/**
+ * @brief This user object serves as the "interface" for interpolating MOOSE variable values and
+ * gradients from the current solution vector.
+ *
+ * Other objects can "couple" to variable values and/or gradients using through this interface.
+ */
+class NEML2FEInterpolation : public ElementUserObject
+{
+public:
+  static InputParameters validParams();
+
+  NEML2FEInterpolation(const InputParameters & parameters);
+
+  /**
+   * @brief Get the variable value of a MOOSE nonlinear variable converted to a NEML2 tensor.
+   *
+   * The NEML2 tensor will have shape (nelem, nqp;), where nelem is the number of elements, and nqp
+   * is the number of quadrature points per element. The tensor is stored in the device memory.
+   *
+   * @warning The tensor value is only available after finalize() is called.
+   */
+  const neml2::Tensor & getValue(const std::string & var_name) const;
+
+  /**
+   * @brief Get the variable gradient of a MOOSE nonlinear variable converted to a NEML2 tensor.
+   *
+   * The NEML2 tensor will have shape (nelem, nqp; 3), where nelem is the number of elements,
+   * and nqp is the number of quadrature points per element. The tensor is stored in the device
+   * memory.
+   *
+   * @warning The tensor value is only available after finalize() is called.
+   */
+  const neml2::Tensor & getGradient(const std::string & var_name) const;
+
+  /**
+   * @brief Get the shape function associated with a MOOSE variable.
+   *
+   * The NEML2 tensor will have shape (nelem, ndofe, nqp;), where nelem is the number of elements,
+   * nqp is the number of quadrature points per element, and ndofe is the number of degrees of
```

Comment:
I'm probably missing this. How does all this work with a mesh that would have tets and hex for example in different subdomains. Do we have an error message to get out of that or does it just work fine?

Reply:
I'd assume we'd use multiple block restricted NEML2Assembly objects etc. for each element type...